src/memtester_parsing.py

```diff
@@ -0,0 +1,189 @@
+# (c) Open Compute Project
+# (c) Google LLC
+# (c) Igor Semenov (goshik92@gmail.com)
+# 
+# Use of this source code is governed by an MIT-style
+# license that can be found in the LICENSE file or at
+# https://opensource.org/licenses/MIT.
+
+from sly import Lexer, Parser
+from dataclasses import dataclass
+from typing import Union, Callable
+import re
+
+@dataclass
+class BadAddress:
+    addr: int
+    left_val: int
+    right_val: int
+    
+@dataclass
+class BadLine:
+    addr: int
+    
+@dataclass
+class OkResult:
+    pass
+
+@dataclass
+class Test:
+    name: str
+    result: Union[OkResult, list[Union[BadAddress, BadLine]]]
+    def passed(self) -> bool:
+        return type(self.result) is OkResult
+    
+@dataclass
+class Loop:
+    index: int
+    tests: list[Test]
+    def failed_tests(self) -> list[Test]:
+        return [t for t in self.tests if not t.passed()]
+
+@dataclass
+class MemtesterCallbacks:
```

Comment:
I guess this is just a matter of preference. Some people prefer OOP-style (overriding), others like functional style (callbacks). For this particular case, both approaches seem equivalent to me.

Reply:
people yes; but in this case python is an OOP-style language and it would be more pythonic to follow that pattern. As a sidenote, this also introduces a lot of cognitive load in the caller context, where you build a `MemtesterCallbacks` object (which is also invalid by construction), and then you piecemeal make it functional

Reply:
I am fine with the OOP style if it is considered preferable. I was about to go and switch to it, but then I realized that the functional style has one important advantage. Functions can have separate lexical environments, whereas methods of a class have to share the same lexical environment. In practical terms this means that with classic overriding, it will be impossible to move one of the "callbacks" to a different scope. Currently, this functionality is not required, but it may turn useful in the future (e. g. if we decide to introduce new steps that share the same memtester executable run). In that sense, I consider the callback approach more general. Any thoughts?

Reply:
i'm not gonna insist on this anymore, just saying that from a pythonic standpoint, this is not generally something you see in codebases. The callback style is very reminiscent of **C**, while python has been from the start an OOP lang.
The scope separation you mention isnt really a huge benefit in my view, nor do I see it as necessary in any future development (or if so, can be done with explicitly shared state inside the **handler** class)